.github/coding-style.md

```diff
@@ -0,0 +1,171 @@
+# Coding style guide
+
+An overview of coding idioms and practiced style in stackrox/stackrox.
+
+Whenever you encounter during code review (either as a reviewer or as a PR
+author) something that you feel is non-obvious or not common sense and does not
+fall into what can be considered common Go style, please add it to this guide.
+Thanks!
+
+## On consistency
+
+There are often many ways to express the same thing (and not all of them are
+correct). Consistency in code serves two main purposes:
+
+- Avoiding errors of ignorance.
+- Minimizing reader's surprise and hence facilitate reviews and maintenance.
+
+When you deviate from a common pattern, consider leaving a comment explaining
+why you do so. Prefer local consistency over global if you have to choose.
+
+## On readability
+
+One of the things that significantly impairs code readability is the amount of
+non-local reasoning required to understand the contract and assumptions. Strive
+to design your types and interfaces so that they can only be used in the
+expected, correct way. 
+```go
+// Bad: do all `cache` users, current and future know how to use it correctly?  
+var (
+	cache = NewCache()
+)
+...
+func (w *widget) Update(x X) error {
+    w.mutex.Lock()
+	defer w.mutex.Unlock()
+	
+	if cache.Contains(x) {
+	    ...	
+    }
+	...
+}
+```
+
+```go
+// Better: `cache` users likely understand that manipulations require
+// synchronization might miss future changes to this contract. 
+var (
+	cache = NewCache()
+    mutex sync.RWMutex
+)
+...
+func (w *widget) Update(x X) error {
+    mutex.Lock()
+	defer mutex.Unlock()
+	
+	if cache.Contains(x) {
+	    ...	
+    }
+	...
+}
+```
+
+```go
+// Good: `cache` is only used by `widget` methods likely living in the same file.
+type widget struct {
+  cache Cache
+  mutex sync.RWMutex
+}
+...
+func (w *widget) Update(x X) error {
+    w.mutex.Lock()
+	defer w.mutex.Unlock()
+	
+	if w.cache.Contains(x) {
+	    ...	
+    }
+	...
+}
+```
+
+Capture in comments assumptions and decisions that are not obvious from the
+code, for example:
+```go
+var (
+	authorizer = perrpc.FromMap(map[authz.Authorizer][]string{
+		// When this endpoint was public, Sensor relied on it to check Central's
+		// availability. While Sensor might not do so today, we need to ensure
+		// backward compatibility with older Sensors.
+		or.SensorOr(user.Authenticated()): {
+			"/v1.MetadataService/GetMetadata",
+		},
+	})
+)
+```
+
+## Guidelines
+
+- Always use `defer mutex.Unlock()` instead of explicitly calling `Unlock()`.
```

Comment:
I don't agree with the definitive `Always` here. I'd prefer this to be a `Prefer`.

In case you'd have something like
```go
func doSomething() {
        mutex.Lock()
        defer mutex.Unlock()
        
        // Read something under lock.
       vals := data["some-key"]

       // Do some heavy operation with `vals`
       calculate(vals)
}
```
We should unlock _after_ we read from the map, and not hold the lock unnecessarily long. While in most cases it _should_ be `defer mutex.Unlock()`, there's an argument to not have this in some places.

Reply:
I don't think this is the pattern we currently have in the codebase and we probably should not introduce it. For your example above, this is how the code should look like:
```
func doSomething() {
	synchronizedF := func() Typename {
		mutex.Lock()
		defer mutex.Unlock()

		// Read something under lock.
		vals := data["some-key"]
		return vals
	}

	vals := synchronizedF()
	
	// Do some heavy operation with `vals`
	calculate(vals)
}
```

Reply:
@rukletsov Realistically the code should look like 

```
func getVals() ValType {
  mutex.Lock()
  defer mutex.Unlock()

  return data["some-key"]
}

func doSomething() { 
  vals := getVals()
  // Do some heavy operation with `vals` 
  calculate(vals)
}
```

@dhaus67 We do have some primitives around concurrency.WithLock when you are trying to do things that very small and another function feels like overkill. A good example of this is:
```
concurrency.WithLock(&d.deploymentProcessingLock, func() {
    delete(d.deploymentProcessingMap, alertResults.GetDeploymentId())
  })
```

Reply:
I would actually like to add some generic forms of WithLock that return values. Could be something like WithLock1 and WithLock2 that return 1 and 2 values respectively so you could basically not have to write the above function. But as it stands, the above function is very explicit. @dhaus67 one problem with not using defer is when panics happen the lock is not always relinquished. An example is this: 
```
func TestPanic(t *testing.T) {
	var lock sync.Mutex

	// m is never initialized
	var m map[string]string

	safe.Run(func() {
		lock.Lock()
		panic("i did something that panics")
		m["adding to nil map"] = "oh no"
		lock.Unlock()
	})
	// Try to get lock, but this will hang forever
	lock.Lock()
}
```

concurrency.WithLock never has this problem